dvc/repo/__init__.py

```diff
@@ -454,14 +455,23 @@ def is_dvc_internal(self, path):
     @contextmanager
     def open(self, path, remote=None, mode="r", encoding=None):
         """Opens a specified resource as a file descriptor"""
+        if not remote and self.tree.exists(path):
```

Comment:
`tree.exists()` will be `True` if someone done `checkout` and might not be what we need if dvc file then changed (e.g. by git checkout). Shouldn't we first look in cached outputs and then fallback to tree check?

Reply:
Hmm, at the beginning I agreed with @Suor s comment [here](https://github.com/iterative/dvc/pull/2566#pullrequestreview-297395101)
but now that you mention checkout, it probably better to check cache first.

Reply:
@pared Good point. Or maybe we can improve this current approach so that it doesn't suffer from that issue?

Reply:
I guess we could solve that either by checking a checksum, or prompting the user?
Checking the checksum would result in losing the advantage suggested by @Suor.

Reply:
So we should more strictly go over all the possibilities here:
- get `out` at the start
- if we have `out` and it's cached then we go to `._open_cached()`
- if we don't have out or have out, which is not cached, we go `.tree.open()`, no need to check for exists then.

There is never such thing as retrying tree open after cached open or vice versa.

Reply:
That makes sense. I should have started with separating finding out from opening it. @efiop do you have any concerns about this approach?

Reply:
@pared sounds good to me.